### ISSUE TITLE
Setting Max TTL to null should omit it

### ISSUE THREAD

**Issue by jonathonbattista** [hashicorp/terraform-provider-google#16200]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request.
* Please do not leave _+1_ or _me too_ comments, they generate extra noise for issue followers and do not help prioritize the request.
* If you are interested in working on this issue or have submitted a pull request, please leave a comment.
* If an issue is assigned to the `modular-magician` user, it is either in the process of being autogenerated, or is planned to be autogenerated soon. If an issue is assigned to a user, that user is claiming responsibility for the issue. If an issue is assigned to `hashibot`, a community member has claimed the issue already.

### Terraform Version

Terraform `v1.0.0`
on linux_amd64
+ provider registry.terraform.io/hashicorp/google v5.1.0

### Affected Resource(s)

* `google_network_services_edge_cache_service`

### Terraform Configuration Files

```tf
resource "google_network_services_edge_cache_service" "default" {
  name        = var.name
  description = "${var.name} service"

  project = var.project_id

  require_tls           = var.require_tls
  edge_ssl_certificates = var.edge_ssl_certificates

  log_config {
    enable      = var.log_config.enable
    sample_rate = var.log_config.sample_rate
  }

  routing {
    dynamic "host_rule" {
      for_each = var.host_rules
      content {
        description  = host_rule.value.description
        hosts        = host_rule.value.hosts
        path_matcher = "${var.name}-${host_rule.value.path_matcher}-route"
      }
    }

    dynamic "path_matcher" {
      for_each = var.path_matchers

      content {
        name = "${var.name}-${path_matcher.value.name}-route"

        dynamic "route_rule" {
          for_each = path_matcher.value.route_rule

          content {
            description = "The ${var.name} ${route_rule.value.name} route"
            priority    = route_rule.value.priority

            match_rule {
              prefix_match        = route_rule.value.prefix_match
              path_template_match = route_rule.value.path_template_match
              full_path_match     = route_rule.value.full_path_match
            }

            origin = google_network_services_edge_cache_origin.default[path_matcher.key].name

            dynamic "route_action" {
              for_each = route_rule.value.route_action

              content {
                cdn_policy {
                  cache_mode  = route_action.value.cdn_policy.cache_mode
                  default_ttl = route_action.value.cdn_policy.default_ttl
                  client_ttl  = route_action.value.cdn_policy.client_ttl
                  max_ttl = route_action.value.cdn_policy.max_ttl
                  cache_key_policy {
                    include_protocol          = route_action.value.cdn_policy.cache_key_policy.include_protocol
                    exclude_host              = route_action.value.cdn_policy.cache_key_policy.exclude_host
                    included_query_parameters = route_action.value.cdn_policy.cache_key_policy.included_query_parameters
                    included_header_names     = route_action.value.cdn_policy.cache_key_policy.included_header_names
                    included_cookie_names     = route_action.value.cdn_policy.cache_key_policy.included_cookie_names
                  }
                  negative_caching        = route_action.value.cdn_policy.negative_caching
                  signed_request_mode     = route_action.value.cdn_policy.signed_request_mode
                  negative_caching_policy = route_action.value.cdn_policy.negative_caching_policy
                }
                url_rewrite {
                  path_prefix_rewrite   = route_action.value.url_rewrite.path_prefix_rewrite
                  host_rewrite          = route_action.value.url_rewrite.host_rewrite
                  path_template_rewrite = route_action.value.url_rewrite.path_template_rewrite
                }
              }
            }

            header_action {
              dynamic "response_header_to_add" {
                for_each = route_rule.value.response_header_to_add
                content {
                  header_name  = response_header_to_add.value.header_name
                  header_value = response_header_to_add.value.header_value
                }
              }

              dynamic "response_header_to_remove" {
                for_each = route_rule.value.response_header_to_remove
                content {
                  header_name = response_header_to_remove.value.header_name
                }
              }

              dynamic "request_header_to_add" {
                for_each = route_rule.value.request_header_to_add
                content {
                  header_name  = request_header_to_add.value.header_name
                  header_value = request_header_to_add.value.header_value
                }
              }

              dynamic "request_header_to_remove" {
                for_each = route_rule.value.request_header_to_remove
                content {
                  header_name = request_header_to_remove.value.header_name
                }
              }
            }
          }
        }
      }
    }
  }
}
```

### Debug Output

https://gist.github.com/jonathonbattista/6afd4748b8df77e06e321fb9b69594b7

### Expected Behavior

Since `max_ttl` is `null`, it should not be considered set and passed the GCP API

### Actual Behavior

`googleapi: Error 400: code:3 message:"Config validation failed" details:{[type.googleapis.com/google.rpc.BadRequest]:{field_violations:{field:"resource.routing.path_matchers[0].route_rules[0].route_action.cdn_policy.max_ttl" description:"maxTtl must not be set when cacheMode is USE_ORIGIN_HEADERS, FORCE_CACHE_ALL, or BYPASS_CACHE."}}}`

### Steps to Reproduce

1. Set `cache_mode` to `FORCE_CACHE_ALL` for a Route Action's `cdn_policy`
2. Set `max_ttl` to `null` for a Route Action's `cdn_policy`
3. `terraform apply`

**Comment by edwardmedia** [comment]
@jonathonbattista do you have the log that show more api requests and responses before the error was thrown out?

**Comment by edwardmedia** [comment]
@jonathonbattista does removing `max_ttl` work for you?

**Comment by jonathonbattista** [comment]
So I had a `path_matchers` variable configured to require each argument for the `route_rules` and `route_actions`.

Initially, we weren't force caching anything so the `max_ttl` argument wasn't an issue, but when we added additional route actions it became an issue.

However, removing the `max_ttl` from the variable structure seems to work as the module doesn't care if it does or doesn't exist, only the API.

TLDR;  omitting the `max_ttl` argument or explicity passing `max_ttl = null` should be interpreted the same way

**Comment by edwardmedia** [comment]
@roaks3 do you have any better suggestions?
